docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-Vasyl-Teliman.md

```diff
@@ -0,0 +1,307 @@
+
+Mesa issues: [#3418](https://gitlab.freedesktop.org/mesa/mesa/-/issues/3418), [#3427](https://gitlab.freedesktop.org/mesa/mesa/-/issues/3427),
+[#3428](https://gitlab.freedesktop.org/mesa/mesa/-/issues/3428), [#3452](https://gitlab.freedesktop.org/mesa/mesa/-/issues/3452),
+[#3453](https://gitlab.freedesktop.org/mesa/mesa/-/issues/3453).
+
+spirv-opt issues: [#3643](https://github.com/KhronosGroup/SPIRV-Tools/issues/3643), [#3738](https://github.com/KhronosGroup/SPIRV-Tools/issues/3738),
+[#3631](https://github.com/KhronosGroup/SPIRV-Tools/issues/3631),
+[#3715](https://github.com/KhronosGroup/SPIRV-Tools/issues/3715), [#3708](https://github.com/KhronosGroup/SPIRV-Tools/issues/3708),
+[#3707](https://github.com/KhronosGroup/SPIRV-Tools/issues/3707), [#3706](https://github.com/KhronosGroup/SPIRV-Tools/issues/3706),
+[#3635](https://github.com/KhronosGroup/SPIRV-Tools/issues/3635), [#3650](https://github.com/KhronosGroup/SPIRV-Tools/issues/3650),
+[#3515](https://github.com/KhronosGroup/SPIRV-Tools/issues/3515).
+
+SPIR-V specification and spirv-val: [#3716](https://github.com/KhronosGroup/SPIRV-Tools/issues/3716), [#3650](https://github.com/KhronosGroup/SPIRV-Tools/issues/3650).
+
+#### Example
+One of the implemented transformations was very helpful in uncovering a bug in the Mesa open-source library. The transformation propagates an instruction from its
+basic block into the block's predecessors. This transformation is a part of a larger idea of being able to permute instructions in the function. Concretely, this transformation:
+1. Copies some instruction from its basic block into the block's predecessors
+2. Replaces the original instruction with an `OpPhi` instruction that selects one of the copies based on the previously executed block
+
+In SPIR-V-like pseudocode:
+<table>
+  <tr>
+    <th>
+      Reference shader
+    </th>
+    <th>
+      Variant shader
+    </th>
+  </tr>
+  <tr>
+    <td>
+      <pre>
+        <code>
+%1 = OpLabel
+     OpBranch %3
+%2 = OpLabel
+     OpBranch %3
+%3 = OpLabel
+<strong>%4 = OpIAdd %int %a %b</strong>
+...
+</code>
+      </pre>
+    </td>
+    <td>
+      <pre>
+        <code>
+%1 = OpLabel
+<strong>%5 = OpIAdd %int %a %b</strong>
+     OpBranch %3
+%2 = OpLabel
+<strong>%6 = OpIAdd %int %a %b</strong>
+     OpBranch %3
+%3 = OpLabel
+<strong>%4 = OpPhi %int %5 %1 %6 %2</strong>
+...
+</code>
+      </pre>
+    </td>
+  </tr>
+</table>
+This transformation triggers a bug in Mesa open-source library that causes the shader to generate a wrong image. 
+
+Reference shader output | Variant shader output
+--- | ---
+![Reference shader output](https://imgur.com/LPQxiRr.png) | ![Variant output shader](https://imgur.com/o383jfp.png)
+
+The bug is triggered when the transformation moves increment and comparison instructions from the loop's back-edge block into its header as follows.
```

Comment:
"as follows." -> "as follows:"

Also, here you say that the transformation moves things from the back-edge block into the header, but then after the code snippet you say that an instruction is propagated from the header block into its predecessors. Can you make these consistent?

Also could you show before and after for this example?  That would make it a lot clearer.

(I think we should include this example in a paper on spirv-fuzz - it's really nice!)

Reply:
Fixed in 8d484e3.